azure_agent.go

```diff
@@ -0,0 +1,210 @@
+// (c) Copyright IBM Corp. 2022
+
+package instana
+
+import (
+	"bytes"
+	"context"
+	"encoding/json"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"net/http"
+	"os"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/instana/go-sensor/acceptor"
+	"github.com/instana/go-sensor/autoprofile"
+)
+
+const (
+	flushPeriodInSec = 2
+
+	azureCustomRuntime string = "custom"
+)
+
+type azureAgent struct {
+	Endpoint string
+	Key      string
+	PID      int
+
+	snapshot serverlessSnapshot
+
+	mu        sync.Mutex
+	spanQueue []Span
+
+	client *http.Client
+	logger LeveledLogger
+}
+
+func newAzureAgent(acceptorEndpoint, agentKey string, client *http.Client, logger LeveledLogger) *azureAgent {
+	if logger == nil {
+		logger = defaultLogger
+	}
+
+	if client == nil {
+		client = http.DefaultClient
+		client.Timeout = 500 * time.Millisecond
+	}
+
+	logger.Debug("initializing azure agent")
+
+	agent := &azureAgent{
+		Endpoint: acceptorEndpoint,
+		Key:      agentKey,
+		PID:      os.Getpid(),
+		client:   client,
+		logger:   logger,
+	}
+
+	go func() {
+		t := time.NewTicker(flushPeriodInSec * time.Second)
+		defer t.Stop()
+
+		for range t.C {
+			if err := agent.Flush(context.Background()); err != nil {
+				agent.logger.Error("failed to post collected data: ", err)
+			}
+		}
+	}()
+
+	return agent
+}
+
+func (a *azureAgent) Ready() bool { return true }
+
+func (a *azureAgent) SendMetrics(acceptor.Metrics) error { return nil }
+
+func (a *azureAgent) SendEvent(*EventData) error { return nil }
+
+func (a *azureAgent) SendSpans(spans []Span) error {
+	a.enqueueSpans(spans)
+	return nil
+}
+
+func (a *azureAgent) SendProfiles([]autoprofile.Profile) error { return nil }
+
+func (a *azureAgent) Flush(ctx context.Context) error {
+	a.collectSnapshot()
+
+	if a.snapshot.EntityID == "" {
+		return ErrAgentNotReady
+	}
+
+	from := newServerlessAgentFromS(a.snapshot.EntityID, "azure")
+
+	payload := struct {
+		Metrics metricsPayload `json:"metrics,omitempty"`
+		Spans   []Span         `json:"spans,omitempty"`
+	}{
+		Metrics: metricsPayload{
+			Plugins: []acceptor.PluginPayload{
+				acceptor.NewAzurePluginPayload(a.snapshot.EntityID),
+			},
+		},
+	}
+
+	a.mu.Lock()
+	payload.Spans = make([]Span, len(a.spanQueue))
+	copy(payload.Spans, a.spanQueue)
+	a.spanQueue = a.spanQueue[:0]
+	a.mu.Unlock()
+
+	for i := range payload.Spans {
+		payload.Spans[i].From = from
+	}
+
+	buf := bytes.NewBuffer(nil)
+	if err := json.NewEncoder(buf).Encode(payload); err != nil {
+		return fmt.Errorf("failed to marshal traces payload: %s", err)
+	}
+
+	payloadSize := buf.Len()
+	if payloadSize > maxContentLength {
+		a.logger.Warn(fmt.Sprintf("failed to send the spans. Payload size: %d exceeded max size: %d", payloadSize, maxContentLength))
+		return payloadTooLargeErr
```

Comment:
Same thing here: should we add a debug to log that spans will be lost?
We don't send them back to the buffer of spans. Which is valid, since the size will remain the same in the next flush attempt, and it would stop here again anyway.